### DISCUSSION TITLE
More Secure Database Access

### DISCUSSION THREAD

**Discussion by uvmskier** [teslamate-org/teslamate#4003] [Ideas]
The same database user is used by teslamate to write to the Postgres database and to connect from Grafana.  A more secure approach would be to split this into two user accounts.  The current user account could be the account used to connect Teslamate to Postgres which needs read/write access.  A new user account could be created to connect Grafana with select only privileges.  This would reduce the possibility that someone accidentally modifies their data and create a more secure environment if you want to give someone access to the Grafana UI but not write access to Postgres.

**Comment by brianmay** [comment]
I have been thinking this for a while, but no time to followup.

Actually I did do something like this for my own. Then I forgot about it.

Something like:

```
create user grafana with password 'xxxx';
GRANT CONNECT ON DATABASE teslamate to grafana;
GRANT USAGE ON SCHEMA teslamate to grafana;
GRANT USAGE ON SCHEMA public to grafana;
GRANT SELECT ON ALL TABLES IN SCHEMA public TO grafana;
ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT ON TABLES TO grafana;
```

I am not sure why I listed the teslamate schema there, I might have been confused. I can only see the public schema now.

ideally grafana shouldn't have any access to the `tokens` table either.
